File: python/add_two_numbers.py
"""
Problem: Add Two Numbers
Difficulty: Medium
Date: 2025-09-15
Language: Python
Approach: Iterative with carry
Time: O(max(m, n))
Space: O(max(m, n))
"""

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
    
    def __repr__(self):
        """Helper method to display linked list"""
        result = []
        current = self
        while current:
            result.append(str(current.val))
            current = current.next
        return "[" + ",".join(result) + "]"

def addTwoNumbers(l1, l2):
    """
    Add two numbers represented as linked lists where digits are stored in reverse order.
    
    Args:
        l1 (ListNode): First number as linked list
        l2 (ListNode): Second number as linked list
    
    Returns:
        ListNode: Sum as linked list in reverse order
    """
    dummy = ListNode(0)
    current = dummy
    carry = 0
    
    while l1 or l2 or carry:
        # Get values from current nodes (0 if node is None)
        val1 = l1.val if l1 else 0
        val2 = l2.val if l2 else 0
        
        # Calculate sum and new carry
        total = val1 + val2 + carry
        carry = total // 10
        digit = total % 10
        
        # Create new node with the digit
        current.next = ListNode(digit)
        current = current.next
        
        # Move to next nodes if they exist
        l1 = l1.next if l1 else None
        l2 = l2.next if l2 else None
    
    return dummy.next

def create_linked_list(digits):
    """Helper function to create linked list from list of digits"""
    if not digits:
        return None
    
    head = ListNode(digits[0])
    current = head
    for digit in digits[1:]:
        current.next = ListNode(digit)
        current = current.next
    
    return head

def linked_list_to_list(head):
    """Helper function to convert linked list to list for easy comparison"""
    result = []
    current = head
    while current:
        result.append(current.val)
        current = current.next
    return result

if __name__ == "__main__":
    # Test cases
    test_cases = [
        ([2, 4, 3], [5, 6, 4], [7, 0, 8]),  # 342 + 465 = 807
        ([0], [0], [0]),                      # 0 + 0 = 0
        ([9, 9, 9, 9, 9, 9, 9], [9, 9, 9, 9], [8, 9, 9, 9, 0, 0, 0, 1])  # 9999999 + 9999 = 10009998
    ]
    
    for i, (l1_digits, l2_digits, expected) in enumerate(test_cases):
        l1 = create_linked_list(l1_digits)
        l2 = create_linked_list(l2_digits)
        result_head = addTwoNumbers(l1, l2)
        result = linked_list_to_list(result_head)
        
        print(f"Test {i+1}: {'PASS' if result == expected else 'FAIL'}")
        print(f"  Input: l1={l1_digits}, l2={l2_digits}")
        print(f"  Expected: {expected}")
        print(f"  Got: {result}")
        print()

File: rust/add_two_numbers.rs
/*
Problem: Add Two Numbers
Difficulty: Medium
Date: 2025-09-15
Language: Rust
Approach: Iterative with carry
Time: O(max(m, n))
Space: O(max(m, n))
*/

// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode {
            next: None,
            val,
        }
    }
}

impl Solution {
    /// Add two numbers represented as linked lists where digits are stored in reverse order.
    /// 
    /// # Arguments
    /// * `l1` - First number as linked list
    /// * `l2` - Second number as linked list
    /// 
    /// # Returns
    /// Sum as linked list in reverse order
    pub fn add_two_numbers(
        l1: Option<Box<ListNode>>,
        l2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        let mut dummy = Box::new(ListNode::new(0));
        let mut current = &mut dummy;
        let mut carry = 0;
        let mut l1 = l1;
        let mut l2 = l2;
        
        while l1.is_some() || l2.is_some() || carry > 0 {
            // Get values from current nodes (0 if node is None)
            let val1 = l1.as_ref().map_or(0, |node| node.val);
            let val2 = l2.as_ref().map_or(0, |node| node.val);
            
            // Calculate sum and new carry
            let total = val1 + val2 + carry;
            carry = total / 10;
            let digit = total % 10;
            
            // Create new node with the digit
            current.next = Some(Box::new(ListNode::new(digit)));
            current = current.next.as_mut().unwrap();
            
            // Move to next nodes if they exist
            l1 = l1.and_then(|node| node.next);
            l2 = l2.and_then(|node| node.next);
        }
        
        dummy.next
    }
}

pub struct Solution;

/// Helper function to create linked list from vector of digits
fn create_linked_list(digits: Vec<i32>) -> Option<Box<ListNode>> {
    if digits.is_empty() {
        return None;
    }
    
    let mut head = Box::new(ListNode::new(digits[0]));
    let mut current = &mut head;
    
    for &digit in digits.iter().skip(1) {
        current.next = Some(Box::new(ListNode::new(digit)));
        current = current.next.as_mut().unwrap();
    }
    
    Some(head)
}

/// Helper function to convert linked list to vector for easy comparison
fn linked_list_to_vec(head: Option<Box<ListNode>>) -> Vec<i32> {
    let mut result = Vec::new();
    let mut current = head;
    
    while let Some(node) = current {
        result.push(node.val);
        current = node.next;
    }
    
    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_add_two_numbers() {
        let test_cases = vec![
            (vec![2, 4, 3], vec![5, 6, 4], vec![7, 0, 8]), // 342 + 465 = 807
            (vec![0], vec![0], vec![0]),                     // 0 + 0 = 0
            (vec![9, 9, 9, 9, 9, 9, 9], vec![9, 9, 9, 9], vec![8, 9, 9, 9, 0, 0, 0, 1]), // 9999999 + 9999 = 10009998
        ];

        for (i, (l1_digits, l2_digits, expected)) in test_cases.iter().enumerate() {
            let l1 = create_linked_list(l1_digits.clone());
            let l2 = create_linked_list(l2_digits.clone());
            let result_head = Solution::add_two_numbers(l1, l2);
            let result = linked_list_to_vec(result_head);
            
            assert_eq!(result, *expected, "Test case {} failed", i + 1);
            println!("Test {}: PASS", i + 1);
            println!("  Input: l1={:?}, l2={:?}", l1_digits, l2_digits);
            println!("  Expected: {:?}", expected);
            println!("  Got: {:?}", result);
            println!();
        }
    }
}

fn main() {
    println!("Add Two Numbers - Rust Implementation");
    println!("=====================================");
    println!();
    
    let test_cases = vec![
        (vec![2, 4, 3], vec![5, 6, 4], vec![7, 0, 8]), // 342 + 465 = 807
        (vec![0], vec![0], vec![0]),                     // 0 + 0 = 0
        (vec![9, 9, 9, 9, 9, 9, 9], vec![9, 9, 9, 9], vec![8, 9, 9, 9, 0, 0, 0, 1]), // 9999999 + 9999 = 10009998
    ];

    for (i, (l1_digits, l2_digits, expected)) in test_cases.iter().enumerate() {
        let l1 = create_linked_list(l1_digits.clone());
        let l2 = create_linked_list(l2_digits.clone());
        let result_head = Solution::add_two_numbers(l1, l2);
        let result = linked_list_to_vec(result_head);
        
        let passed = result == *expected;
        println!("Test {}: {}", i + 1, if passed { "PASS" } else { "FAIL" });
        println!("  Input: l1={:?}, l2={:?}", l1_digits, l2_digits);
        println!("  Expected: {:?}", expected);
        println!("  Got: {:?}", result);
        println!();
    }
    
    println!("Run tests with: cargo test");
}
